pkg/spoke/hubclientcert/controller.go

```
@@ -0,0 +1,217 @@
+	if hasValidKubeconfig(secret) {
+		commonName := fmt.Sprintf("%s%s:%s", subjectPrefix, c.clusterName, c.agentName)
+		leaf, err := getCertLeaf(secret, commonName)
+		if err != nil {
+			return err
+		}
+
+		total := leaf.NotAfter.Sub(leaf.NotBefore)
+		remaining := leaf.NotAfter.Sub(time.Now())
+		if remaining.Seconds()/total.Seconds() > 0.2 {
+			return nil
+		}
+		klog.Infof("Ther current client certificate for hub expires in %v. Start certificate rotation", remaining.Round(time.Second))
+	} else {
+		klog.Info("No valid client certificate for hub is found. Bootstrap is required")
+	}
+
+	// create a csr
+	c.keyData, err = keyutil.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return err
+	}
+
+	c.csrName, err = createCSR(c.hubCSRClient, c.keyData, c.clusterName, c.agentName)
+	if err != nil {
+		return err
+	}
+	klog.Infof("A csr %q is created", c.csrName)
+	return nil
+}
+
+func (c *ClientCertForHubController) syncCSR(secret *corev1.Secret) (bool, error) {
+	// skip if there is no ongoing csr
+	if c.csrName == "" {
+		return false, nil
+	}
+
+	// skip if csr no longer exists
+	csr, err := c.hubCSRLister.Get(c.csrName)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			c.csrName = ""
+			return false, nil
+		}
+		return false, err
+	}
+
+	// skip if csr is not approved yet
+	if !isCSRApproved(csr) {
+		return false, nil
+	}
+
+	// skip if csr has no certificate in its status yet
+	if csr.Status.Certificate == nil {
+		return false, nil
+	}
+
+	klog.V(4).Infof("Sync csr %q", c.csrName)
+	// check if cert in csr status matches with the corresponding private key
+	if c.keyData == nil {
+		klog.Warningf("No private key found for certificate in csr: %s", c.csrName)
+		c.csrName = ""
+		return false, nil
+	}
+	_, err = tls.X509KeyPair(csr.Status.Certificate, c.keyData)
+	if err != nil {
+		klog.Warningf("Private key does not match with the certificate in csr: %s", c.csrName)
+		c.csrName = ""
+		return false, nil
```

return an error instead of just having a warning.  it makes event, it shows in metrics for requeues/errors, and can be reflected in status.